### PR TITLE
src: add log API

### PR DESCRIPTION
**Pull request by trevnorris**
Add a C++ API to collect logs written with the nsolid JS API.

Log support for Winston and Pino will also be added.

### CODE REVIEW

File: src/nsolid.h [hunk 1]
```diff
@@ -583,6 +589,18 @@ NODE_EXTERN int OnUnblockedLoopHook(Cb&& cb, Data&&... data);
 template <typename Cb, typename... Data>
 NODE_EXTERN int OnConfigurationHook(Cb&& cb, Data&&... data);
 
+/**
+ * @brief Regsiter a hook(function) to be called when a log line is written.
```
Comment (santigimeno) [inline src/nsolid.h]:
```suggestion
 * @brief Register a hook(function) to be called when a log line is written.
```

File: src/nsolid/nsolid_api.h [hunk 2]
```diff
@@ -634,6 +649,9 @@ class EnvList {
   // strings that will be passed to the user's hooks.
   TSList<OnConfigHookStor> on_config_hook_list_;
   TSQueue<std::string> on_config_string_q_;
+  // List for OnLogWriteHook callbacks.
+  TSList<OnLogWriteHookStor> on_log_write_hook_list_;
+  TSQueue<std::pair<SharedEnvInst, std::string>> on_log_write_q_;
```
Comment (santigimeno) [inline src/nsolid/nsolid_api.h]: We'll need to use a specific `struct` to store the associated metadata (log_level, etc.)

File: src/nsolid.h [hunk 3]
```diff
@@ -1764,6 +1805,13 @@ void on_configuration_hook_proxy_(std::string info, void* data) {
   (*static_cast<G*>(data))(info);
 }
 
+template <typename G>
+void on_log_write_hook_proxy_(SharedEnvInst inst,
+                              std::string line,
+                              void* data) {
+  (*static_cast<G*>(data))(inst, line);
```
Comment (santigimeno) [inline src/nsolid.h]: Maybe use `std::move() here. I'm not too fond of those 2 string copies per hook call.
```suggestion
  (*static_cast<G*>(data))(inst, std::move(line));
```

**Review by santigimeno — CHANGES_REQUESTED** [review]

File: lib/nsolid.js [hunk 4]
```diff
@@ -212,6 +219,11 @@ start.traceStats = assignGetters({}, {
 module.exports = start;
 
 
+function writeLog(msg, sev) {
+  binding.writeLog(typeof msg === 'string' ? msg : JSONStringify(msg), sev);
+}
```
Comment (santigimeno) [inline lib/nsolid.js:224]: Should we check other types as well such as Buffer? I don't really know tbh

File: src/nsolid.h [hunk 5]
```diff
@@ -583,6 +590,27 @@ NODE_EXTERN int OnUnblockedLoopHook(Cb&& cb, Data&&... data);
...
 NODE_EXTERN int OnConfigurationHook(Cb&& cb, Data&&... data);
 
+/**
+ * @brief The struct containing log info that's passed to the OnLogWriteHook
+ * callback.
+ */
+struct LogWriteInfo {
+  std::string info;
+  int level;
+};
```
Comment (santigimeno) [inline src/nsolid.h]: By looking at https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#log-and-event-record-definition, at the very least we should support:
- Timestamp
- TraceId / SpanId / TraceFlags if tracing is enabled.

We may add more later if we thing we need them.

**Review by santigimeno — APPROVED** [review]

File: lib/nsolid.js (hunk 4 above)
Comment (trevnorris) [reply lib/nsolid.js:224]: That's a good idea, but I wasn't sure if we should match how `console.log()` works on output.

File: src/nsolid.h [hunk 6]
```diff
@@ -1764,6 +1815,13 @@ void on_configuration_hook_proxy_(std::string info, void* data) {
   (*static_cast<G*>(data))(info);
 }
 
+template <typename G>
+void on_log_write_hook_proxy_(SharedEnvInst inst,
+                              LogWriteInfo info,
+                              void* data) {
+  (*static_cast<G*>(data))(inst, info);
```
Comment (santigimeno) [inline src/nsolid.h]: I forgot to add... do you think it's possible to avoid making 2 copies of `LogWriteInfo` on every log?

File: src/nsolid.h [hunk 7]
```diff
@@ -1770,6 +1825,13 @@ void on_configuration_hook_proxy_(std::string info, void* data) {
   (*static_cast<G*>(data))(info);
 }
 
+template <typename G>
+void on_log_write_hook_proxy_(SharedEnvInst inst,
+                              LogWriteInfo info,
+                              void* data) {
+  (*static_cast<G*>(data))(inst, std::move(info));
+}
```
Comment (santigimeno) [inline src/nsolid.h:1833]: Does the compiler know how to move LogWriteInfo or do we have to implement the move constructor ourselves?

**Review by santigimeno — APPROVED** [review]

File: src/nsolid.h (hunk 7 above)
Comment (trevnorris) [reply src/nsolid.h:1833]: `LogWriteInfo` contains only standard types that it knows how to move. So we should be fine.

**Pull request merged**